File: Machine_Learning_A_Z/Part 2 - Regression/Section 6 - Polynomial Regression/1 - PolinomialLinearRegression.py
#POLINOMIAL LINEAR REGRESSION

#IMPORTING LIBRARIES
import numpy as np 
import matplotlib.pyplot as plt
import pandas as pd

#IMPORTING THE DATASET
dataset = pd.read_csv('Position_Salaries.csv')

#CREATING HTE MATRIX OF FEATURES
x = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values



#We wont split the dataset because we do not have enough data to split into train and test

#------------------------------------FITTING SIMPLE LINEAR REGRESSION TO DATASET------------------------------------#
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(x, y)



#------------------------------------FITTING POLYNOMIAL LINEAR REGRESSION TO DATASET------------------------------------#
from sklearn.preprocessing import PolynomialFeatures
#poly_reg = PolynomialFeatures(degree = 1)
#poly_reg = PolynomialFeatures(degree = 2)
#poly_reg = PolynomialFeatures(degree = 3)
poly_reg = PolynomialFeatures(degree = 4)
x_poly = poly_reg.fit_transform(x)

lin_reg_2 = LinearRegression()
lin_reg_2.fit(x_poly, y)



#------------------------------------VISUALIZATION------------------------------------#
#Simple Linear Regression
plt.scatter(x, y, color = "red")
plt.plot(x, lin_reg.predict(x))
plt.title("Truth Plot (Simple Regression)")
plt.xlabel("Level")
plt.ylabel("Salary")

#Polynomial linear regression
plt.scatter(x, y, color = "red")
#plt.plot(x, lin_reg_2.predict(x_poly))
#or
plt.plot(x, lin_reg_2.predict(poly_reg.fit_transform(x)))
plt.title("Truth Plot (Polynomial Regression)")
plt.xlabel("Level")
plt.ylabel("Salary")

#improving the polynomial regression above but creating increments of 0.1
#for a better curve
x_grid = np.arange(min(x), max(x), 0.1)
x_grid = x_grid.reshape(len(x_grid), 1)
plt.scatter(x, y, color = "red")
plt.plot(x_grid, lin_reg_2.predict(poly_reg.fit_transform(x_grid)))
plt.title("Truth Plot (Polynomial Regression)")
plt.xlabel("Level")
plt.ylabel("Salary")



#------------------------------------PREDICTING A RESULT------------------------------------#
#Predicting for simple linear regression
lin_reg.predict(6.5)

#Predicting for Polynomial linear regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))

File: Machine_Learning_A_Z/Part 2 - Regression/Section 6 - Polynomial Regression/1 - PolinomialLinearRegression.R
#POLINOMIAL LINEAR REGRESSION

dataset <- read.csv('Position_Salaries.csv')
dataset <- dataset[, 2:3]
#Dataset is too small to split into test and train datasets



#------------------------------------FITTING SIMPLE LINEAR REGRESSION TO DATASET------------------------------------#
lin_reg <- lm(Salary ~., data = dataset)
summary(lin_reg)



#------------------------------------FITTING POLYNOMIAL LINEAR REGRESSION TO DATASET------------------------------------#
dataset$Level2 <- dataset$Level^2
dataset$Level3 <- dataset$Level^3
dataset$Level4 <- dataset$Level^4
poly_reg <- lm(Salary ~., data = dataset)
summary(poly_reg)


#------------------------------------VISUALIZATION------------------------------------#
#Simple Linear Regression
library(ggplot2)
ggplot() + geom_point(aes(dataset$Level, dataset$Salary), color = "RED") +
  geom_line(aes(dataset$Level, predict(lin_reg,dataset)), color = "BLUE") +
  ggtitle("Truth Plot (Linear Regression)") +
  xlab("Years of Experience") +
  ylab("Salary")


#Polynomial Linear Regression
library(ggplot2)
ggplot() + geom_point(aes(dataset$Level, dataset$Salary), color = "RED") +
  geom_line(aes(dataset$Level, predict(poly_reg,dataset)), color = "BLUE") +
  ggtitle("Truth Plot (Polynomial Regression)") +
  xlab("Years of Experience") +
  ylab("Salary")



#------------------------------------PREDICTING A RESULT------------------------------------#
#Predicting for simple linear regression
predict(lin_reg, data.frame(Level = 6.5))

#Predicting for Polynomial linear regression
predict(poly_reg, data.frame(Level = 6.5, Level2 = 6.5^2, Level3 = 6.5^3, Level4 = 6.5^4))
